decoder.go

```diff
@@ -0,0 +1,211 @@
+	// {type: { structFieldTag:  }}
+	structSpecs = make(map[reflect.Type]map[string]*structField)
+	mut         sync.RWMutex
+)
+
+// Scan scans the results from a key-value Redis map result set ([]interface{})
+// to a destination struct. The Redis keys are matched to the struct's field
+// tags defined in the struct indifcated by fieldTag.
+// eg: fieldTag=redis where Name string `redis:"name"`
+func Scan(vals []interface{}, dest interface{}, fieldTag string) error {
+	if len(vals)%2 != 0 {
+		return errors.New("args should have an even number of items (key-val)")
+	}
+
+	// The destination to scan into should be a struct pointer.
+	ob := reflect.ValueOf(dest)
+	if ob.Kind() != reflect.Ptr || ob.IsNil() {
+		return errors.New("dest must be a non-nil pointer to a struct")
+	}
+	ob = ob.Elem()
+
+	if ob.Kind() != reflect.Struct {
+		return errors.New("dest must be a non-nil pointer to a struct")
+	}
+
+	// If the struct field spec is not cached, build and cache it to avoid
+	// iterating through the fields of a struct type every time values are
+	// scanned into it.
+	typ := ob.Type()
+	mut.RLock()
+	fMap, ok := structSpecs[typ]
+	mut.RUnlock()
+
+	if !ok {
+		fMap = makeStructSpecs(ob, fieldTag)
+		mut.Lock()
+		structSpecs[typ] = fMap
+		mut.Unlock()
+	}
+
+	// Iterate through the (key, value) sequence.
+	for i := 0; i < len(vals); i += 2 {
+		key, ok := vals[i].(string)
+		if !ok {
+			continue
+		}
+
+		val, ok := vals[i+1].(string)
+		if !ok {
+			continue
+		}
+
+		// Check if the field name is in the field spec map.
+		field, ok := fMap[key]
+		if !ok {
+			continue
+		}
+
+		// Invoke the custom decoder or the default decoder that's registered
+		// on the field in the spec map.
+		switch fn := field.fn.(type) {
+		case Decoder:
+			if err := fn.Decode(val); err != nil {
+				return err
+			}
+		case defDecoder:
+			if err := fn(ob.Field(field.index), val); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func makeStructSpecs(ob reflect.Value, fieldTag string) map[string]*structField {
+	var (
+		num = ob.NumField()
+		out = make(map[string]*structField)
+	)
+
+	for i := 0; i < num; i++ {
+		f := ob.Field(i)
+		if !f.IsValid() || !f.CanSet() {
+			continue
+		}
+
+		tag := ob.Type().Field(i).Tag.Get(fieldTag)
+		if tag == "" || tag == "-" {
+			continue
+		}
+
+		tag = strings.Split(tag, ",")[0]
+		if tag == "" {
+			continue
+		}
+
+		// If the field has a custom Decode() (implements Decoder) defined on it, use that.
+		if df, ok := f.Addr().Interface().(Decoder); ok {
+			out[tag] = &structField{index: i, fn: df}
+			continue
+		}
+
+		// If there is a default decoder defined for the type, use that.
+		if df, ok := defDecoders[f.Kind()]; ok {
+			out[tag] = &structField{index: i, fn: df}
+			continue
+		}
+	}
+
+	return out
+}
+
+func decodeBool(f reflect.Value, s string) error {
+	b, err := strconv.ParseBool(s)
+	if err != nil {
+		return errors.New("expected boolean")
```

Comment:
Please don't hide the original error here and below. It should be just `return err`.